parlai/agents/language_model/language_model.py

```diff
@@ -0,0 +1,469 @@
+        shared = super().share()
+        shared['dict'] = self.dict
+        shared['NULL_IDX'] = self.NULL_IDX
+        shared['END_IDX'] = self.END_IDX
+        if self.opt.get('numthreads', 1) > 1:
+            shared['model'] = self.model
+            self.model.share_memory()
+            shared['states'] = self.states
+        return shared
+
+    def observe(self, observation):
+        """Save observation for act.
+        If multiple observations are from the same episode, concatenate them.
+        """
+        #shallow copy observation (deep copy can be expensive)
+        obs = observation.copy()
+        seq_len = self.opt['seq_len']
+        is_training = True
+        if 'eval_labels' in obs:
+            is_training = False
+
+        if is_training:
+            if 'text' in obs:
+                vec = self.parse(obs['text'])
+                vec.append(self.END_IDX)
+                self.next_observe += vec
+            if 'labels' in obs:
+                vec = self.parse(obs['labels'][0])
+                vec.append(self.END_IDX)
+                self.next_observe += vec
+            if len(self.next_observe) < (seq_len + 1):
+                # not enough to return to make a batch
+                # we handle this case in vectorize
+                # labels indicates that we are training
+                self.observation = {'labels': ''}
+                return self.observation
+            else:
+                vecs_to_return = []
+                total = len(self.next_observe) // (seq_len + 1)
+                for _ in range(total):
+                    observe = self.next_observe[:(seq_len + 1)]
+                    self.next_observe = self.next_observe[(seq_len + 1):]
+                    vecs_to_return.append(observe)
+                dict_to_return = {'text': '', 'labels': '', 'text2vec': vecs_to_return}
+                self.observation = dict_to_return
+                return dict_to_return
+        else:
+            self.observation = obs
+            return obs
+
+    def repackage_hidden(self, h):
+        """Wraps hidden states in new Variables, to detach them from their history."""
+        if type(h) == Variable:
+            return Variable(h.data)
+        else:
+            return tuple(self.repackage_hidden(v) for v in h)
+
+    def get_target_loss(self, data, hidden, targets, y_lens):
+        """Calculates the loss with respect to the targets, token by token,
+           where each output token is conditioned on either the input or the
+           previous target token.
+        """
+        loss = 0.0
+        bsz = data.size(0)
+
+        # feed in inputs without end token
+        output, hidden = self.model(data.transpose(0,1), hidden)
+        self.hidden = self.repackage_hidden(hidden)
+        # feed in end token
+        ends = Variable(torch.LongTensor([self.END_IDX for _ in range(bsz)]).view(1, bsz))
+        if self.use_cuda:
+            ends = ends.cuda()
```

Comment:
one last thing here--I think it would be better to just initialize this Tensor once (e.g. in __init__, including pushing it to the GPU if that's set) and then you can pack it in a variable every time here instead.
e.g. https://github.com/facebookresearch/ParlAI/blob/master/parlai/agents/seq2seq/modules.py#L64

Reply:
i'm actually confused .. are you commenting on things rewritten by @emilydinan  -- or were these in the original code she started from?

Reply:
I thought of doing that, but the problem is that "bsz" isn't always the actual batchsize. Most of the time it is, but we get rid of examples that are not "valid." I guess it will be better to initialize it once, since it is used most of the time.

Reply:
my comment is on the current code above. re: actual bsz, you can use `expand` to size it to the current batch